### ISSUE TITLE
Parallel Workers to load models at one end point to server parallel users

### ISSUE THREAD

**Issue by eivorwolfkissed** [oobabooga/text-generation-webui#5732]
The current api end point is really great. I am able to load the model in multiple end points and serve parallel users. Could you add the feature to load the model in parallel at the same end point, so that the concurrent users could be routed to available models.
Something like "from vllm.engine.async_llm_engine import AsyncLLMEngine"

**Comment by realroz** [comment]
You could spin up several workers and use a load balancer to spread out the load. That is what I'm doing. Just know that if you have two workers on the same machine, the inference speed is very slow, so you need many machines / GPUs.

**Comment by eivorwolfkissed** [comment]
Thank you for the reply. This is what we do today, on the servers with multile GPUs, I load the model 'n' number of times at different endpoints and load balance. If i go for smaller models, that means i will have to load balance between a lot of endpoints (say 100).
Could you let me know the load balance approach you follow to check if an endpoint is available for inference ?
vllm has an interesting approach. There a model can be loaded 'n' number of times in one endpoint and this gives the option to have different models for different purpose at different end points.

**Comment by TiagoTiago** [comment]
I'm sorry if this is off-topic, but it kinda sounds like something I've been looking for. Would this allow me to load one big model with some layers offloaded to the GPU, and one small model all on the CPU? Is this already possible in the current version? How would I set that up?

**Comment by eivorwolfkissed** [comment]
Yes, with the current set up it is possible to have what you are asking for at different endpoints. But scaling becomes challenging.  What i am looking for is multi workers at one endpoint.

**Comment by TiagoTiago** [comment]
> Yes, with the current set up it is possible to have what you are asking for at different endpoints. But scaling becomes challenging. What i am looking for is multi workers at one endpoint.

Ah, so this feature request is what would allow me to load two models on the same machine simultaneously, and as it is right now that is not yet possible?

**Comment by TiagoTiago** [comment]
Since you seem to know what you're talking about, and this kinda sounds like it could be straying away from the exact topic of the feature request, can you come help me on the [discussion thread](https://github.com/oobabooga/text-generation-webui/discussions/5667) where I asked about this recently please?

**Comment by realroz** [comment]
So, how I do it may change in the future. I have yet to deploy at scale. I load as many LLMs as my machines can handle. Then I use NGINX to load balance.  Here is my localhost load balancer: 
``` 
http {
    upstream backend {
        server host.docker.internal:5000;
        server host.docker.internal:5001;
    }

    server {
        listen 8080;

        location /v1/chat/completions {
            if ($request_method = 'OPTIONS') {
                add_header 'Access-Control-Allow-Origin' "$http_origin";
                add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS';
                add_header 'Access-Control-Allow-Headers' 'Content-Type, Accept';
                add_header 'Content-Length' '0';
                add_header 'Content-Type' 'text/plain charset=UTF-8';
                return 204;
            }

            proxy_pass http://backend/v1/chat/completions;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
            proxy_set_header Content-Type "application/json";
            proxy_set_header Accept "application/json";
            add_header Access-Control-Allow-Origin "$http_origin" always;
            add_header 'Access-Control-Allow-Headers' 'Content-Type, Accept';
        }
    }
}
events {}

```

In this instance, I'll spin up two text gen UIs on my localhost (macos), and my load balancer runs in docker. You can add more URLS (as many as you have to the upstream backend). Then, the single URL you use is your load balancer. This will spread the load out. The more workers you have, the more likely a worker will be idle and ready for inference. But you only need to advertise the URL of the load balancer.

Feel free to let me know if you have any questions or if you see opportunities for me to improve my approach. Thank you!

**Comment by miaoyizi** [comment]
> 您可以启动多个工作线程并使用负载均衡器来分散负载。这就是我正在做的事情。只要知道，如果你在同一台机器上有两个工作人员，推理速度会非常慢，因此你需要很多机器/GPU。

你好，这个也是我想要做的，解决并发API请求的问题。 一个模型上是否支持多线程进行并发请求处理？  或者部署多个节点，再加上一个负载均衡，

**Comment by miaoyizi** [comment]
> I'm sorry if this is off-topic, but it kinda sounds like something I've been looking for. Would this allow me to load one big model with some layers offloaded to the GPU, and one small model all on the CPU? Is this already possible in the current version? How would I set that up?

Hello, this is what I want to do, solve the problem of concurrent API requests. Does a model support multithreading for concurrent request processing? The second option is to deploy multiple nodes, preceded by a load balancing that requires more hardware
